### PR TITLE
COMP: add completion for cargo features in cfg attributes

### PR DESCRIPTION
**Pull request by vlad20012**
Peek 2020-11-25 18-55

Related to #4693

changelog: Provide completion for cargo features in `#[cfg()]` attributes

### CODE REVIEW

File: toml/src/main/kotlin/org/rust/toml/completion/RsCargoTomlIntegrationCompletionContributor.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+/** Provides completion in **Rust** files for elements that points to TOML elements, e.g. for cargo features */
+class RsCargoTomlIntegrationCompletionContributor : CompletionContributor() {
+    init {
+        if (tomlPluginIsAbiCompatible()) {
+            extend(BASIC, RsCfgFeatureCompletionProvider)
+        }
+    }
+
+    private fun extend(type: CompletionType?, provider: RsCompletionProvider) {
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/completion/RsCargoTomlIntegrationCompletionContributor.kt:22]: Let's add `@Suppress("SameParameterValue")` or inline this function

File: toml/src/main/kotlin/org/rust/toml/completion/RsCfgFeatureCompletionProvider.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+private fun rustLookupElementForFeature(feature: TomlKey): LookupElementBuilder {
+    return LookupElementBuilder
+        .createWithSmartPointer(feature.text, feature)
+        .withInsertHandler(RustStringLiteralInsertionHandler())
+}
+
+private class RustStringLiteralInsertionHandler : InsertHandler<LookupElement> {
+    override fun handleInsert(context: InsertionContext, item: LookupElement) {
+        val leaf = context.getElementOfType<PsiElement>() ?: return
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/completion/RsCfgFeatureCompletionProvider.kt:61]: Not sure but this looks a bit more clear IMO
```suggestion
        val leaf = context.getElementOfType<LeafPsiElement>() ?: return
```

**Review by ortem — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/rust/lang/core/completion/RsCfgAttributeCompletionProvider.kt [hunk 3]
```diff
@@ -16,12 +15,14 @@ import com.intellij.patterns.PatternCondition
...
 import com.intellij.util.ProcessingContext
-import org.rust.lang.RsLanguage
+import org.rust.lang.core.psi.RsElementTypes
 import org.rust.lang.core.psi.RsMetaItem
 import org.rust.lang.core.psi.RsMetaItemArgs
+import org.rust.lang.core.psi.RsPath
 import org.rust.lang.core.psi.ext.name
 import org.rust.lang.core.psiElement
 
+/** See also `RsCfgFeatureCompletionProvider` */
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/completion/RsCfgAttributeCompletionProvider.kt:25]: They're in different gradle modules, so KDoc links doesn't wok

File: src/main/kotlin/org/rust/lang/core/completion/RsCfgAttributeCompletionProvider.kt [hunk 4]
```diff
@@ -16,12 +15,14 @@ import com.intellij.patterns.PatternCondition
...
 object RsCfgAttributeCompletionProvider : RsCompletionProvider() {
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/completion/RsCfgAttributeCompletionProvider.kt:26]: Yes! I'm going to do this in a separate PR

File: src/main/kotlin/org/rust/lang/core/completion/RsCfgAttributeCompletionProvider.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ortem) [reply src/main/kotlin/org/rust/lang/core/completion/RsCfgAttributeCompletionProvider.kt:25]: Looks like it works fine in my case (with navigation) using IntelliJ IDEA 2020.3 Beta with Kotlin `203-1.4.10-release-IJ5981.20`

Screenshot 2020-11-26 at 11 21 42

File: toml/src/main/kotlin/org/rust/toml/completion/RsCfgFeatureCompletionProvider.kt (hunk 2 above)
Comment (vlad20012) [reply toml/src/main/kotlin/org/rust/toml/completion/RsCfgFeatureCompletionProvider.kt:61]: The code is mostly copied from TOML's `StringLiteralInsertionHandler`

File: src/main/kotlin/org/rust/lang/core/completion/RsCfgAttributeCompletionProvider.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/completion/RsCfgAttributeCompletionProvider.kt:25]: But if not redered:
image

**Review by ortem — APPROVED** [review]
bors r+

**Comment by vlad20012** [comment]
bors retry

**Comment by vlad20012** [comment]
bors r=ortem

**Pull request merged**